src/main/java/com/yscope/logging/log4j1/ClpIrFileAppender.java

```diff
@@ -0,0 +1,431 @@
+import org.apache.commons.io.output.CountingOutputStream;
+import org.apache.log4j.EnhancedPatternLayout;
+import org.apache.log4j.HTMLLayout;
+import org.apache.log4j.Layout;
+import org.apache.log4j.PatternLayout;
+import org.apache.log4j.SimpleLayout;
+import org.apache.log4j.spi.ErrorCode;
+import org.apache.log4j.spi.LoggingEvent;
+import org.apache.log4j.xml.XMLLayout;
+
+import java.io.*;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.time.ZonedDateTime;
+
+/**
+ * A Log4j appender that writes log events into a Zstandard-compressed CLP IR
+ * stream file.
+ * <p></p>
+ * Since this appender buffers data in the process of compressing the output,
+ * derived appenders should ensure the appender is closed even when the program
+ * exits uncleanly. Otherwise, the compressed output may be truncated. When the
+ * appender is used directly from Log4j, we install a shutdown hook for this
+ * purpose.
+ */
+public class ClpIrFileAppender extends EnhancedAppenderSkeleton implements Flushable {
+  private String timestampPattern = null;
+
+  private int estimatedFormattedTimestampLength = 0;
+  private long uncompressedSizeInBytes = 0;
+
+  private final ByteArrayOutputStream messageWithThrowableBuffer = new ByteArrayOutputStream();
+
+  private AbstractClpIrOutputStream clpIrOutputStream;
+  private CountingOutputStream countingOutputStream;
+
+  private boolean activated = false;
+
+  // Appender settings
+  // NOTE: It may appear that these settings are never set but Log4j sets these
+  // through reflection
+  private int compressionLevel = 3;
+  private boolean closeFrameOnFlush = true;
+  private String filePath;
+  private boolean useFourByteEncoding = false;
+
+  /**
+   * Default constructor (necessary for Log4j to instantiate the appender using
+   * reflection).
+   */
+  public ClpIrFileAppender () {
+  }
+
+  /**
+   * Constructs a ClpIrFileAppender
+   * @param filePath Output file path
+   * @param layout Log4j layout for formatting log events. Only
+   * {@code org.apache.log4j.EnhancedPatternLayout},
+   * {@code org.apache.log4j.PatternLayout}, and
+   * {@code org.apache.log4j.SimpleLayout} are supported.
+   * @param useFourByteEncoding Whether to use the four-byte encoding instead
+   * of the default eight-byte encoding
+   * @param closeFrameOnFlush Whether to close the Zstandard frame on flush
+   * @param compressionLevel Compression level to use for Zstandard
+   * @throws IOException on I/O error
+   */
+  public ClpIrFileAppender (
+      String filePath,
+      Layout layout,
+      boolean useFourByteEncoding,
+      boolean closeFrameOnFlush,
+      int compressionLevel
+  ) throws IOException {
+    super();
+
+    setFile(filePath);
+    setLayout(layout);
+    setUseFourByteEncoding(useFourByteEncoding);
+    setCompressionLevel(compressionLevel);
+    setCloseFrameOnFlush(closeFrameOnFlush);
+
+    super.activateOptions();
```

Comment:
The class defined its own activateOptions method, why not using that one instead of super's?

Reply:
Deduplicated into a helper method that allows us to throw exceptions in this constructor and not in `activateOptions`.